### PR TITLE
Add more explicit nullable types for default null values

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Continuing to bring deprecation-free support for PHP 8.4 (https://php.watch/versions/8.4/implicitly-marking-parameter-type-nullable-deprecated)

If this gets merged, I'll take care of adding missing ones in upper branches if not done during upmerge 🙂

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php [hunk 1]
```diff
@@ -157,7 +157,7 @@ public function __construct(Dunglas $j, Dunglas $k)
 
 class OptionalParameter
 {
-    public function __construct(CollisionInterface $c = null, A $a, Foo $f = null)
+    public function __construct(A $a, ?CollisionInterface $c = null, ?Foo $f = null)
```
Comment (alexandre-daubois) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php]: Swapped because it triggers a deprecation : optional parameter cannot be before mandatory parameter.

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php]: this test case was about optional argument being before mandatory ones, so we'd better keep this and mark `testOptionalParameter` with `@group legacy`

Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php]: Alright, added. However, I was not able to catch the deprecation with the `ExpectDeprecationTrait`, thus leading to have it in the test report. Is there another way to expect it, as it is raised by the language?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php]: Ah right, this should be moved to a standalone fixture file

Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php]: Here you go 👍

**Comment by nicolas-grekas** [comment]
Looks like there are more found by the CI?

**Comment by alexandre-daubois** [comment]
I think the remaining ones are raised by Composer and PHPUnit, which have been fixed in recent commits: https://github.com/composer/composer/commit/62126e1a402c2f5252a1e6026c20765ceae49252 and https://github.com/sebastianbergmann/phpunit/commit/eb22ea1dafbe60991f17fb62522ebf3279903a63

**Comment by nicolas-grekas** [comment]
Thank you @alexandre-daubois.

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php [hunk 2]
```diff
@@ -405,6 +406,9 @@ public function testResolveParameter()
         $this->assertEquals(Foo::class, $container->getDefinition('bar')->getArgument(0));
     }
 
+    /**
+     * @group legacy
```
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:410]: Why was this change necessary?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:410]: This test case is specifically about the deprecated syntax, testing autowiring of arguments with default value positioned before a required one.

Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:410]: To get rid of a wanted deprecation, see https://github.com/symfony/symfony/pull/54308#discussion_r1528733421 for the details 🙂

Comment (xabbuh) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:410]: Understood, but why wasn't the legacy group required before?

Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:410]: I don't know. It is deprecated since 8.1 (https://3v4l.org/n4leq), however I wouldn't be able to explain why this warning pops now 🤔

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:410]: This was not deprecated in 8.1 but has just been in 8.4. It was not deprecated because that's the only syntax that was once supported on older versions to make a type nullable.
https://3v4l.org/K8qWH

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:410]: https://3v4l.org/K8qWH

Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:410]: Thanks for the info 👍
